api/src/main/java/org/openmrs/CohortMembership.java

```diff
@@ -99,6 +120,20 @@ public void setEndDate(Date endDate) {
 
 	@Override
 	public int compareTo(CohortMembership o) {
-		return this.getPatient().getPatientId() - o.getPatient().getPatientId();
+		int ret = -1;
+		if (Objects.equals(this.getPatientId(), o.getPatientId())
+				&& Objects.equals(this.getCohort().getCohortId(), o.getCohort().getCohortId())
+				&& this.getStartDate().equals(o.getStartDate())
+				&& OpenmrsUtil.compare(this.getStartDate(), o.getStartDate()) == 0
+				&& ((this.getEndDate() != null && o.getEndDate() != null
+				&& OpenmrsUtil.compare(this.getEndDate(), o.getEndDate()) == 0)
+				|| (this.getEndDate() == null && o.getEndDate() == null))) {
+			ret = 0;
+		} else if (this.isActive() && !o.isActive()) {
+			ret = -1;
+		} else if (!this.isActive() && o.isActive()) {
+			ret = 1;
+		}
+		return ret;
```

Comment:
_Copying my comment from an older commit:_

`return OpenmrsUtil.compare(Date, Date)` seems to be breaking some tests in CohortTest. 

<img width="572" alt="screen shot 2017-04-14 at 6 05 29 pm" src="https://cloud.githubusercontent.com/assets/193819/25059567/abf17190-213d-11e7-89e3-df328112dd4e.png">

Assert of `cohort.contains(id)` fails sometimes. It can no longer match patient ids. 
```
		Arrays.stream(ids).forEach(id -> assertTrue(cohort.contains(new Patient(id))));
```